src/ripple/nodestore/DeterministicShard.md

```diff
@@ -0,0 +1,103 @@
+# Deterministic Database Shards
+
+This doc describes the standard way to assemble the database shard. A shard assembled using this approach becomes deterministic i.e. if two independent sides assemble the shard consists of the same ledgers, accounts and transactions, then they will obtain the same shard files `nudb.dat` and `nudb.key`. The approach deals with the `NuDB` database format only, refer to `https://github.com/vinniefalco/NuDB`.
+
+
+## Headers
+
+Due to NuDB database definition, the following headers are used for database files:
+
+nudb.key:
+```
+char[8]         Type            The characters "nudb.key"
+uint16          Version         Holds the version number
+uint64          UID             Unique ID generated on creation
+uint64          Appnum          Application defined constant
+uint16          KeySize         Key size in bytes
+uint64          Salt            A random seed
+uint64          Pepper          The salt hashed
+uint16          BlockSize       Size of a file block in bytes
+uint16          LoadFactor      Target fraction in 65536ths
+uint8[56]       Reserved        Zeroes
+uint8[]         Reserved        Zero-pad to block size
+```
+
+nudb.dat:
+```
+char[8]         Type            The characters "nudb.dat"
+uint16          Version         Holds the version number
+uint64          UID             Unique ID generated on creation
+uint64          Appnum          Application defined constant
+uint16          KeySize         Key size in bytes
+uint8[64]       (reserved)      Zeroes
+```
+there all fields are saved using network byte order (most significant byte first).
+
+To make the shard deterministic the following parameters are used as values of header field both for `nudb.key` and `nudb.dat` files.
+```
+Version         2
+UID             digest(0)
+Appnum          digest(2) | 0x5348524400000000 /* 'SHRD' */
+KeySize         32
+Salt            digest(1)
+Pepper          XXH64(Salt)
+BlockSize       0x1000 (4096 bytes)
+LoadFactor      0.5 (numeric 0x8000)
+```
+Note: XXH64() is well-known hash algorithm.
+
+The `digest(i)` mentioned above defined as the follows:
+
+First, RIPEMD160 hash `H` calculated of the following structure
+```
+uint256         lastHash        Hash of last ledger in shard
+uint32          index           Index of the shard
+uint32          firstSeq        Sequence number of first ledger in the shard
+uint32          lastSeq         Sequence number of last ledger in the shard
+uint32          version         Version of shard, 2 at the present
+```
+there all 32-bit integers are hashed in network byte order.
+
+Then, `digest(i)` is defined as the following portion of the above hash `H`:
+```
+digest(0) = H[0] << 56 | H[2] << 48 | ... | H[14] << 0,
```

Comment:
Why all this interleaving again instead of just cutting the RIPEMD160 hash in 3 parts? This sounds needlessly complex, either you trust that this is a cryptographically secure hash (then you don't need this obfuscation) or you don't (then there are other ways to get a 32 and 2 64 bit numbers from 384 bits of input).

Reply:
I documented here the piece of code proposed by Nik Bougalis. RIPEMD160 is also his proposition. Please refer to 
[#3363](https://github.com/ripple/rippled/pull/3363).

Reply:
Done.